From 83318832668ecc90916c145d9c0db2312d831f5e Mon Sep 17 00:00:00 2001
From: "sujuan.chen" <sujuan.chen@mediatek.com>
Date: Wed, 14 Aug 2024 16:23:13 +0800
Subject: [PATCH] muldf: add mtk unify log dump framework muldf-helper

add debug helper to receive log/pkt/cmd/event from driver and send them
to pc-tool/muldfd

Signed-off-by: Bo Jiao <Bo.Jiao@mediatek.com>
Signed-off-by: sujuan.chen <sujuan.chen@mediatek.com>
---
 include/linux/soc/mediatek/muldf.h |  96 +++
 net/Kconfig                        |   2 +
 net/Makefile                       |   1 +
 net/muldf_helper/Kconfig           |   7 +
 net/muldf_helper/Makefile          |   8 +
 net/muldf_helper/debugfs.c         | 780 +++++++++++++++++++++++
 net/muldf_helper/muldf_helper.c    | 977 +++++++++++++++++++++++++++++
 net/muldf_helper/muldf_helper.h    | 214 +++++++
 net/muldf_helper/trace.c           | 124 ++++
 net/muldf_helper/trace.h           | 123 ++++
 net/muldf_helper/vnet.c            |  81 +++
 11 files changed, 2413 insertions(+)
 create mode 100644 include/linux/soc/mediatek/muldf.h
 create mode 100644 net/muldf_helper/Kconfig
 create mode 100644 net/muldf_helper/Makefile
 create mode 100644 net/muldf_helper/debugfs.c
 create mode 100644 net/muldf_helper/muldf_helper.c
 create mode 100644 net/muldf_helper/muldf_helper.h
 create mode 100644 net/muldf_helper/trace.c
 create mode 100644 net/muldf_helper/trace.h
 create mode 100644 net/muldf_helper/vnet.c

diff --git a/include/linux/soc/mediatek/muldf.h b/include/linux/soc/mediatek/muldf.h
new file mode 100644
index 0000000..d8322da
--- /dev/null
+++ b/include/linux/soc/mediatek/muldf.h
@@ -0,0 +1,96 @@
+/* SPDX-License-Identifier: ISC */
+/*
+ * Copyright (C) 2024 MediaTek Inc.
+ */
+#ifndef __MULDF_H
+#define __MULDF_H
+
+#include <linux/kernel.h>
+#include <linux/skbuff.h>
+
+enum muldf_log_level {
+	MTK_MSG_EXCESSIVE,
+	MTK_MSG_MSGDUMP,
+	MTK_MSG_DEBUG,
+	MTK_MSG_INFO,
+	MTK_MSG_WARNING,
+	MTK_MSG_ERROR
+};
+
+enum muldf_dump_type {
+	MTK_DUMP_NONE,
+	MTK_DUMP_PKT,
+	MTK_DUMP_CMD,
+	_MTK_DUMP_MAX
+};
+
+#ifdef CONFIG_MULDF_HELPER
+u64 muldf_alloc_pid(void);
+void muldf_xmit_event(u16 id);
+void muldf_xmit_ics(const u8 *data, u16 len);
+void muldf_xmit_pkt(u64 pid, const u8 *data, u16 len);
+void muldf_xmit_fwlog(const u8 *data, u16 len);
+
+int muldf_add_module(char *module, int def_level);
+int muldf_add_cat(char *module, char *cat);
+int muldf_add_sub_cat(char *module, char *cat, char *sub_cat, u8 pkt_dump);
+
+int muldf_del_module(char *module);
+
+void muldf_print(int level, char *module,
+		 char *cat, char *sub_cat, char *tag,
+		 const char *func, const u32 line,
+		 const char *fmt, ...);
+void muldf_print_hex(int level, char *module,
+		     char *cat, char *sub_cat, char *tag,
+		     u8 *buf, u16 len, const char *title);
+void muldf_print_pkt(int level, char *module,
+		     char *cat, char *sub_cat, char *tag,
+		     u8 *buf, u16 len, const char *fmt, ...);
+#else
+static inline u64 muldf_alloc_pid(void) { return 0; }
+static inline void muldf_xmit_event(u16 id)
+{
+	do {} while (0);
+}
+static inline void muldf_xmit_ics(const u8 *data, u16 len)
+{
+	do {} while (0);
+}
+static inline void muldf_xmit_pkt(u64 pid, const u8 *data, u16 len)
+{
+	do {} while (0);
+}
+static inline void muldf_xmit_fwlog(const u8 *data, u16 len)
+{
+	do {} while (0);
+}
+
+static inline int muldf_add_module(char *module, int def_level) { return -1; }
+static inline int muldf_add_cat(char *module, char *cat) { return -1; }
+static inline int muldf_add_sub_cat(char *module, char *cat, char *sub_cat, u8 pkt_dump)
+{
+	return -1;
+}
+static inline void muldf_print(int level, char *module,
+			       char *cat, char *sub_cat, char *tag,
+			       const char *func, const u32 line,
+			       const char *fmt, ...)
+{
+	do {} while (0);
+}
+static inline void muldf_print_hex(int level, char *module,
+				   char *cat, char *sub_cat, char *tag,
+				   u8 *buf, u16 len, const char *title)
+{
+	do {} while (0);
+}
+static inline void muldf_print_pkt(int level, char *module,
+				   char *cat, char *sub_cat, char *tag,
+				   u8 *buf, u16 len, const char *fmt, ...)
+{
+	do {} while (0);
+}
+#endif
+
+#endif /* __MULDF_H */
diff --git a/net/Kconfig b/net/Kconfig
index f8f6b55..124cce6 100644
--- a/net/Kconfig
+++ b/net/Kconfig
@@ -26,6 +26,8 @@ menuconfig NET
 
 if NET
 
+source "net/muldf_helper/Kconfig"
+
 config ETHERNET_PACKET_MANGLE
 	bool
 	help
diff --git a/net/Makefile b/net/Makefile
index bfde291..624d722 100644
--- a/net/Makefile
+++ b/net/Makefile
@@ -89,3 +89,4 @@ endif
 obj-$(CONFIG_QRTR)		+= qrtr/
 obj-$(CONFIG_NET_NCSI)		+= ncsi/
 obj-$(CONFIG_XDP_SOCKETS)	+= xdp/
+obj-y				+= muldf_helper/
diff --git a/net/muldf_helper/Kconfig b/net/muldf_helper/Kconfig
new file mode 100644
index 0000000..4f755f7
--- /dev/null
+++ b/net/muldf_helper/Kconfig
@@ -0,0 +1,7 @@
+# SPDX-License-Identifier: ISC
+
+config MULDF_HELPER
+	bool "MediaTek Wlan System Unify Log Dump Framework Helper"
+	default n
+	help
+	  If you want to enable Mediatek wlan debug framework, say Y.
diff --git a/net/muldf_helper/Makefile b/net/muldf_helper/Makefile
new file mode 100644
index 0000000..1eef783
--- /dev/null
+++ b/net/muldf_helper/Makefile
@@ -0,0 +1,8 @@
+# SPDX-License-Identifier: ISC
+#
+# Makefile for the Mediatek Debug built-in
+#
+
+obj-$(CONFIG_MULDF_HELPER) := muldf_helper.o
+
+muldf_helper-y := muldf_helper.o trace.o debugfs.o vnet.o
diff --git a/net/muldf_helper/debugfs.c b/net/muldf_helper/debugfs.c
new file mode 100644
index 0000000..b70adb1
--- /dev/null
+++ b/net/muldf_helper/debugfs.c
@@ -0,0 +1,780 @@
+// SPDX-License-Identifier: ISC
+/* Copyright (C) 2024 MediaTek Inc. */
+
+#include "asm-generic/int-ll64.h"
+#include "linux/printk.h"
+#include "muldf_helper.h"
+#include "trace.h"
+
+static int mac_aton(char *addr, u8 *n)
+{
+	int matches = sscanf(addr, "%hhx:%hhx:%hhx:%hhx:%hhx:%hhx",
+			     n, n + 1, n + 2, n + 3, n + 4, n + 5);
+
+	return (matches == ETH_ALEN ? 0 : -EINVAL);
+}
+
+static ssize_t
+muldf_helper_para_read(struct file *file,
+		       char __user *user_buf,
+		       size_t count,
+		       loff_t *ppos)
+{
+	struct muldf_helper *p_muldf = file->private_data;
+	char buf[256] = {};
+	int len = 0, bufsz = 256;
+
+	if (!p_muldf) {
+		pr_err("%s: p_muldf is null\n", __func__);
+		return -EINVAL;
+	}
+
+	if (is_zero_ether_addr(p_muldf->local_mac))
+		muldf_helper_get_local_mac();
+
+	len += scnprintf(buf + len, bufsz - len,
+			 "platform=%s\n", p_muldf->platform);
+	len += scnprintf(buf + len, bufsz - len,
+			 "wlan_driver=%s\n", p_muldf->wlan_driver);
+	len += scnprintf(buf + len, bufsz - len,
+			 "wifi_chip=%s\n", p_muldf->wifi_chip);
+	len += scnprintf(buf + len, bufsz - len,
+			 "ip_dest=%pI4\n", &p_muldf->server_ip);
+	len += scnprintf(buf + len, bufsz - len,
+			 "mac_dest=%pM\n", &p_muldf->server_mac);
+	len += scnprintf(buf + len, bufsz - len,
+			 "port_dest=%hu\n", p_muldf->server_port);
+	len += scnprintf(buf + len, bufsz - len,
+			 "ifname=%s\n", p_muldf->local_ifname);
+	len += scnprintf(buf + len, bufsz - len,
+			 "mode=%d\n", p_muldf->mode);
+	len += scnprintf(buf + len, bufsz - len,
+			 "attach=%d\n", p_muldf->attach);
+	len += scnprintf(buf + len, bufsz - len,
+			 "uuid=%s\n", p_muldf->uuid);
+	len += scnprintf(buf + len, bufsz - len,
+			 "url=%s\n", p_muldf->url);
+	len += scnprintf(buf + len, bufsz - len,
+			 "mac=%pM\n", &p_muldf->local_mac);
+
+	return simple_read_from_buffer(user_buf, count, ppos,
+				       buf, len);
+}
+
+static ssize_t
+muldf_helper_para_write(struct file *file,
+			const char __user *user_buf,
+			size_t count,
+			loff_t *ppos)
+{
+	struct muldf_helper *p_muldf = file->private_data;
+	char b[256], *buf, *p;
+	ssize_t ret = count;
+
+	if (!count)
+		return 0;
+
+	if (count > sizeof(b) - 1)
+		buf =  kmalloc(count + 1, GFP_KERNEL);
+	else
+		buf = b;
+
+	if (copy_from_user(buf, user_buf, count)) {
+		ret = -EFAULT;
+		goto out;
+	}
+
+	p = strchr(buf, '=');
+	if (!p) {
+		ret = -EFAULT;
+		goto out;
+	}
+
+	if (!strncmp(buf, "ip_dest", p - buf)) {
+		p_muldf->server_ip = in_aton(p + 1);
+		goto out;
+	}
+
+	if (!strncmp(buf, "mac_dest", p - buf)) {
+		if (mac_aton(p + 1, p_muldf->server_mac))
+			ret = -EFAULT;
+		goto out;
+	}
+
+	/* if (!strncmp(buf, "port_dest", p - buf)) {
+	 *	u16 val;
+	 *
+	 *	if (sscanf(buf, "port_dest=%hu", &val) == 1)
+	 *		p_muldf->server_port = val;
+	 *	else
+	 *		ret = -EFAULT;
+	 *
+	 *	goto out;
+	 *}
+	 */
+
+	if (!strncmp(buf, "platform", p - buf)) {
+		size_t len = strlen(p + 1) + 1;
+		int i;
+		char *bb = kzalloc(len, GFP_KERNEL);
+
+		if (!bb)
+			return -ENOMEM;
+
+		memset(bb, 0, len);
+		memcpy(bb, p + 1, len - 1);
+		for (i = len; i > 0; i--) {
+			if (bb[i] == '\n')
+				bb[i] = 0;
+		}
+
+		strlcpy(p_muldf->platform, bb, sizeof(p_muldf->platform));
+		kfree(bb);
+
+		goto out;
+	}
+
+	if (!strncmp(buf, "wlan_driver", p - buf)) {
+		size_t len = strlen(p + 1) + 1;
+		int i;
+		char *bb = kzalloc(len, GFP_KERNEL);
+
+		if (!bb)
+			return -ENOMEM;
+
+		memset(bb, 0, len);
+		memcpy(bb, p + 1, len - 1);
+		for (i = len; i > 0; i--) {
+			if (bb[i] == '\n')
+				bb[i] = 0;
+		}
+
+		strlcpy(p_muldf->wlan_driver, bb, sizeof(p_muldf->wlan_driver));
+		kfree(bb);
+
+		goto out;
+	}
+
+	if (!strncmp(buf, "wifi_chip", p - buf)) {
+		size_t len = strlen(p + 1) + 1;
+		int i;
+		char *bb = kzalloc(len, GFP_KERNEL);
+
+		if (!bb)
+			return -ENOMEM;
+
+		memset(bb, 0, len);
+		memcpy(bb, p + 1, len - 1);
+		for (i = len; i > 0; i--) {
+			if (bb[i] == '\n')
+				bb[i] = 0;
+		}
+
+		strlcpy(p_muldf->wifi_chip, bb, sizeof(p_muldf->wifi_chip));
+		kfree(bb);
+
+		goto out;
+	}
+
+	if (!strncmp(buf, "if_name", p - buf)) {
+		size_t len = strlen(p + 1) + 1;
+		int i;
+		char *bb = kzalloc(len, GFP_KERNEL);
+
+		if (!bb)
+			return -ENOMEM;
+
+		memset(bb, 0, len);
+		memcpy(bb, p + 1, len - 1);
+		for (i = len; i > 0; i--) {
+			if (bb[i] == '\n')
+				bb[i] = 0;
+		}
+
+		strlcpy(p_muldf->local_ifname, bb, sizeof(p_muldf->local_ifname));
+		kfree(bb);
+
+		goto out;
+	}
+
+	if (!strncmp(buf, "uuid", p - buf)) {
+		size_t len = strlen(p + 1) + 1;
+		int i;
+		char *bb = kzalloc(len, GFP_KERNEL);
+
+		if (!bb)
+			return -ENOMEM;
+
+		memset(bb, 0, len);
+		memcpy(bb, p + 1, len - 1);
+		for (i = len; i > 0; i--) {
+			if (bb[i] == '\n')
+				bb[i] = 0;
+		}
+
+		strlcpy(p_muldf->uuid, bb, sizeof(p_muldf->uuid));
+		kfree(bb);
+
+		goto out;
+	}
+
+	if (!strncmp(buf, "url", p - buf)) {
+		size_t len = strlen(p + 1) + 1;
+		int i;
+		char *bb = kzalloc(len, GFP_KERNEL);
+
+		if (!bb)
+			return -ENOMEM;
+
+		memset(bb, 0, len);
+		memcpy(bb, p + 1, len - 1);
+		for (i = len; i > 0; i--) {
+			if (bb[i] == '\n')
+				bb[i] = 0;
+		}
+
+		strlcpy(p_muldf->url, bb, sizeof(p_muldf->url));
+		kfree(bb);
+
+		goto out;
+	}
+
+	if (!strncmp(buf, "mac", p - buf)) {
+		size_t len = strlen(p + 1) + 1;
+		int i;
+		char *bb = kzalloc(len, GFP_KERNEL);
+
+		if (!bb)
+			return -ENOMEM;
+
+		memset(bb, 0, len);
+		memcpy(bb, p + 1, len - 1);
+		for (i = len; i > 0; i--) {
+			if (bb[i] == '\n')
+				bb[i] = 0;
+		}
+
+		strlcpy(p_muldf->local_mac, bb, sizeof(p_muldf->local_mac));
+		kfree(bb);
+
+		goto out;
+	}
+
+	if (!strncmp(buf, "mode", p - buf)) {
+		u8 val, mode = p_muldf->mode;
+
+		if (sscanf(buf, "mode=%hhu", &val) == 1)
+			mode = val;
+		else
+			ret = -EFAULT;
+
+		//mode changed
+		if (mode != p_muldf->mode) {
+			p_muldf->mode = MTK_MODE_INTERNAL;
+
+			if (mode == MTK_MODE_INTERNAL)
+				muldf_xmit_event(EVENT_ID_MODE_INTERNAL);
+			else
+				muldf_xmit_event(EVENT_ID_MODE_PC);
+
+			p_muldf->mode = mode;
+		}
+
+		goto out;
+	}
+
+	if (!strncmp(buf, "attach", p - buf)) {
+		u8 val;
+
+		if (sscanf(buf, "attach=%hhu", &val) == 1)
+			p_muldf->attach = !!val;
+		else
+			ret = -EFAULT;
+
+		goto out;
+	}
+
+	if (!strncmp(buf, "enable", p - buf)) {
+		u8 val;
+
+		if (sscanf(buf, "enable=%hhu", &val) == 1)
+			p_muldf->enable = val;
+		else
+			ret = -EFAULT;
+
+		if (!p_muldf->enable)
+			muldf_helper_del_event();
+		goto out;
+	}
+
+	if (!strncmp(buf, "event_id", p - buf)) {
+		if (p_muldf->enable) {
+			u16 val;
+
+			if (sscanf(buf, "event_id=%hu", &val) == 1)
+				muldf_helper_add_event(val);
+			else
+				ret = -EFAULT;
+		}
+
+		goto out;
+	}
+
+	if (!strncmp(buf, "module", p - buf)) {
+		u8 module;
+
+		if (sscanf(buf, "module=%hhu", &module) == 1) {
+			//write module trigger module info upload.
+			muldf_helper_get_module_info();
+		} else {
+			ret = -EFAULT;
+		}
+
+		goto out;
+	}
+
+out:
+	if (b != buf)
+		kfree(buf);
+
+	return ret;
+}
+
+static const struct file_operations fops_uldf_para = {
+	.read = muldf_helper_para_read,
+	.write = muldf_helper_para_write,
+	.open = simple_open,
+};
+
+static ssize_t
+muldf_helper_status(struct file *file,
+		    char __user *user_buf,
+		    size_t count,
+		    loff_t *ppos)
+{
+	struct muldf_helper *p_muldf = file->private_data;
+	char buf[256] = {};
+	int len = 0, bufsz = 256;
+
+	if (!p_muldf) {
+		pr_err("%s: p_muldf is null\n", __func__);
+		return -EINVAL;
+	}
+
+	len += scnprintf(buf + len, bufsz - len, "Event:\n");
+	len += scnprintf(buf + len, bufsz - len,
+			 "    event entry max=%d\n", EVENT_ID_MAX_NUM);
+	len += scnprintf(buf + len, bufsz - len,
+			 "    event entry num=%d\n", p_muldf->event_entry_num);
+	len += scnprintf(buf + len, bufsz - len,
+			 "    event queue max len=%d\n", MAX_EVENT_CNT);
+	len += scnprintf(buf + len, bufsz - len,
+			 "    event queue len=%d\n", p_muldf->event_queue.cnt);
+	len += scnprintf(buf + len, bufsz - len,
+			 "    event recv cnt=%d\n", p_muldf->event_queue.total_cnt);
+
+	return simple_read_from_buffer(user_buf, count, ppos,
+				       buf, len);
+}
+
+static const struct file_operations fops_uldf_status = {
+	.read = muldf_helper_status,
+	.write = NULL,
+	.open = simple_open,
+};
+
+static ssize_t
+muldf_helper_event(struct file *file, char __user *buf, size_t count, loff_t *ppos)
+{
+	struct muldf_event_hdr ev;
+	int ret;
+
+	if (*ppos > 0)
+		return 0;
+
+	ret = wait_event_interruptible(p_muldf->event_queue_wait,
+				       atomic_read(&p_muldf->event_queue_status) != 0);
+	if (ret)
+		return -ERESTARTSYS;
+
+	if (muldf_helper_deq_event(&ev) < 0)
+		return -EAGAIN;
+
+	if (copy_to_user(buf, &ev, sizeof(ev)))
+		return -EFAULT;
+
+	*ppos += sizeof(ev);
+	return sizeof(ev);
+}
+
+static const struct file_operations fops_uldf_event = {
+	.read = muldf_helper_event,
+	.write = NULL,
+	.open = simple_open,
+};
+
+static ssize_t
+muldf_helper_uuid(struct file *file, char __user *user_buf, size_t count, loff_t *ppos)
+{
+	struct muldf_helper *p_muldf = file->private_data;
+	char buf[256] = {};
+	int len = 0;
+
+	if (!p_muldf) {
+		pr_err("%s: p_muldf is null\n", __func__);
+		return -EINVAL;
+	}
+
+	len = scnprintf(buf, sizeof(buf), "%s", p_muldf->uuid);
+
+	return simple_read_from_buffer(user_buf, count, ppos,
+				       buf, len);
+}
+
+static const struct file_operations fops_uldf_uuid = {
+	.read = muldf_helper_uuid,
+	.write = NULL,
+	.open = simple_open,
+};
+
+static ssize_t
+muldf_helper_url(struct file *file, char __user *user_buf, size_t count, loff_t *ppos)
+{
+	struct muldf_helper *p_muldf = file->private_data;
+	char buf[256] = {};
+	int len = 0;
+
+	if (!p_muldf) {
+		pr_err("%s: p_muldf is null\n", __func__);
+		return -EINVAL;
+	}
+
+	len = scnprintf(buf, sizeof(buf), "%s", p_muldf->url);
+
+	return simple_read_from_buffer(user_buf, count, ppos,
+				       buf, len);
+}
+
+static const struct file_operations fops_uldf_url = {
+	.read = muldf_helper_url,
+	.write = NULL,
+	.open = simple_open,
+};
+
+static ssize_t
+muldf_helper_mac(struct file *file, char __user *user_buf, size_t count, loff_t *ppos)
+{
+	struct muldf_helper *p_muldf = file->private_data;
+	char buf[256] = {};
+	int len = 0;
+
+	if (!p_muldf) {
+		pr_err("%s: p_muldf is null\n", __func__);
+		return -EINVAL;
+	}
+
+	if (is_zero_ether_addr(p_muldf->local_mac))
+		muldf_helper_get_local_mac();
+
+	len = scnprintf(buf, sizeof(buf), "%pM", p_muldf->local_mac);
+
+	return simple_read_from_buffer(user_buf, count, ppos,
+				       buf, len);
+}
+
+static const struct file_operations fops_uldf_mac = {
+	.read = muldf_helper_mac,
+	.write = NULL,
+	.open = simple_open,
+};
+
+static ssize_t
+muldf_helper_mode(struct file *file, char __user *user_buf, size_t count, loff_t *ppos)
+{
+	struct muldf_helper *p_muldf = file->private_data;
+	char buf[256] = {};
+	int len = 0;
+
+	if (!p_muldf) {
+		pr_err("%s: p_muldf is null\n", __func__);
+		return -EINVAL;
+	}
+
+	len = scnprintf(buf, sizeof(buf), "%d", p_muldf->mode);
+
+	return simple_read_from_buffer(user_buf, count, ppos,
+				       buf, len);
+}
+
+static const struct file_operations fops_uldf_mode = {
+	.read = muldf_helper_mode,
+	.write = NULL,
+	.open = simple_open,
+};
+
+static ssize_t
+muldf_helper_ip(struct file *file, char __user *user_buf, size_t count, loff_t *ppos)
+{
+	struct muldf_helper *p_muldf = file->private_data;
+	char buf[256] = {};
+	int len = 0;
+
+	if (!p_muldf) {
+		pr_err("%s: p_muldf is null\n", __func__);
+		return -EINVAL;
+	}
+
+	len = scnprintf(buf, sizeof(buf), "%pI4", &p_muldf->server_ip);
+
+	return simple_read_from_buffer(user_buf, count, ppos,
+				       buf, len);
+}
+
+static const struct file_operations fops_uldf_ip = {
+	.read = muldf_helper_ip,
+	.write = NULL,
+	.open = simple_open,
+};
+
+static ssize_t
+muldf_helper_platform(struct file *file, char __user *user_buf, size_t count, loff_t *ppos)
+{
+	struct muldf_helper *p_muldf = file->private_data;
+	char buf[32] = {};
+	int len = 0;
+
+	if (!p_muldf) {
+		pr_err("%s: p_muldf is null\n", __func__);
+		return -EINVAL;
+	}
+
+	len = scnprintf(buf, sizeof(buf), "%s", p_muldf->platform);
+
+	return simple_read_from_buffer(user_buf, count, ppos,
+				       buf, len);
+}
+
+static const struct file_operations fops_uldf_platform = {
+	.read = muldf_helper_platform,
+	.write = NULL,
+	.open = simple_open,
+};
+
+static ssize_t
+muldf_helper_driver(struct file *file, char __user *user_buf, size_t count, loff_t *ppos)
+{
+	struct muldf_helper *p_muldf = file->private_data;
+	char buf[32] = {};
+	int len = 0;
+
+	if (!p_muldf) {
+		pr_err("%s: p_muldf is null\n", __func__);
+		return -EINVAL;
+	}
+
+	len = scnprintf(buf, sizeof(buf), "%s", p_muldf->wlan_driver);
+
+	return simple_read_from_buffer(user_buf, count, ppos,
+				       buf, len);
+}
+
+static const struct file_operations fops_uldf_driver = {
+	.read = muldf_helper_driver,
+	.write = NULL,
+	.open = simple_open,
+};
+
+static ssize_t
+muldf_helper_chip(struct file *file, char __user *user_buf, size_t count, loff_t *ppos)
+{
+	struct muldf_helper *p_muldf = file->private_data;
+	char buf[32] = {};
+	int len = 0;
+
+	if (!p_muldf) {
+		pr_err("%s: p_muldf is null\n", __func__);
+		return -EINVAL;
+	}
+
+	len = scnprintf(buf, sizeof(buf), "%s", p_muldf->wifi_chip);
+
+	return simple_read_from_buffer(user_buf, count, ppos,
+				       buf, len);
+}
+
+static const struct file_operations fops_uldf_chip = {
+	.read = muldf_helper_chip,
+	.write = NULL,
+	.open = simple_open,
+};
+
+static ssize_t
+muldf_helper_para_module(struct file *file,
+			 char __user *user_buf,
+			 size_t count,
+			 loff_t *ppos)
+{
+	struct muldf_helper *p_muldf = file->private_data;
+	struct muldf_module *p_mod;
+	struct muldf_cat *p_cat;
+	struct muldf_sub_cat *p_sub_cat;
+	int ret, len = 0, bufsz = 4096;
+	char *buf;
+
+	if (!p_muldf) {
+		pr_err("%s: p_muldf is null\n", __func__);
+		return -EINVAL;
+	}
+
+	buf = kmalloc(bufsz, GFP_KERNEL);
+	if (!buf)
+		return -EINVAL;
+
+	len += scnprintf(buf + len, bufsz - len,
+			 "modules num=%d:\n", p_muldf->module_num);
+
+	rcu_read_lock();
+	list_for_each_entry(p_mod, &p_muldf->module_head, list) {
+		len += scnprintf(buf + len, bufsz - len,
+				 "module=%s \tcategory num=%d:\n", p_mod->name, p_mod->cat_num);
+
+		list_for_each_entry(p_cat, &p_mod->cat_head, list) {
+			len += scnprintf(buf + len, bufsz - len,
+					 "module=%s \tcategory=%s \tsub-category num=%d:\n",
+					 p_mod->name, p_cat->name, p_cat->sub_cat_num);
+
+			list_for_each_entry(p_sub_cat, &p_cat->sub_cat_head, list) {
+				len += scnprintf(buf + len, bufsz - len,
+						"module=%s \tcategory=%s \tsub-category=%s \t[dump_type=%d,dump_en=%d]\n",
+						p_mod->name, p_cat->name, p_sub_cat->name,
+						p_sub_cat->pkt_dump, p_sub_cat->pkt_dump_en);
+			}
+		}
+	}
+	rcu_read_unlock();
+
+	ret = simple_read_from_buffer(user_buf, count, ppos,
+				      buf, len);
+
+	kfree(buf);
+	return ret;
+}
+
+static const struct file_operations fops_muldf_module = {
+	.read = muldf_helper_para_module,
+	.write = NULL,
+	.open = simple_open,
+};
+
+static ssize_t
+muldf_helper_para_dump_write(struct file *file,
+			     const char __user *user_buf,
+			     size_t count,
+			     loff_t *ppos)
+{
+	char buf[100];
+	char module[32];
+	char cat[32];
+	char sub_cat[32];
+	u8 dump_en;
+
+	if (count >= sizeof(buf))
+		return -EINVAL;
+
+	if (copy_from_user(buf, user_buf, count))
+		return -EFAULT;
+
+	if (count && buf[count - 1] == '\n')
+		buf[count - 1] = '\0';
+	else
+		buf[count] = '\0';
+
+	if (sscanf(buf, "%s %s %s %hhu", module, cat, sub_cat, &dump_en) != 4)
+		return -EINVAL;
+
+	muldf_helper_set_module_cat_dump(module, cat, sub_cat, dump_en);
+
+	return count;
+}
+
+static const struct file_operations fops_muldf_dump = {
+	.write = muldf_helper_para_dump_write,
+	.read = NULL,
+	.open = simple_open,
+	.llseek = default_llseek,
+};
+
+static ssize_t
+muldf_helper_para_level_write(struct file *file,
+			      const char __user *user_buf,
+			      size_t count,
+			      loff_t *ppos)
+{
+	char buf[100];
+	char module[32];
+	u8 level;
+
+	if (count >= sizeof(buf))
+		return -EINVAL;
+
+	if (copy_from_user(buf, user_buf, count))
+		return -EFAULT;
+
+	if (count && buf[count - 1] == '\n')
+		buf[count - 1] = '\0';
+	else
+		buf[count] = '\0';
+
+	if (sscanf(buf, "%s %hhu", module, &level) != 2)
+		return -EINVAL;
+
+	muldf_helper_set_module_level(module, level);
+
+	return count;
+}
+
+static const struct file_operations fops_muldf_level = {
+	.write = muldf_helper_para_level_write,
+	.read = NULL,
+	.open = simple_open,
+	.llseek = default_llseek,
+};
+
+int muldf_helper_init_debugfs(struct muldf_helper *p_muldf)
+{
+	struct dentry *dir;
+
+	dir = debugfs_create_dir("muldf_helper", NULL);
+	if (!dir)
+		return -ENOMEM;
+
+	debugfs_create_file("muldf_para", 0600, dir, p_muldf,
+			    &fops_uldf_para);
+	debugfs_create_file("muldf_status", 0600, dir, p_muldf,
+			    &fops_uldf_status);
+	debugfs_create_file("muldf_event", 0600, dir, p_muldf,
+			    &fops_uldf_event);
+	debugfs_create_file("muldf_uuid", 0600, dir, p_muldf,
+			    &fops_uldf_uuid);
+	debugfs_create_file("muldf_url", 0600, dir, p_muldf,
+			    &fops_uldf_url);
+	debugfs_create_file("muldf_mac", 0600, dir, p_muldf,
+			    &fops_uldf_mac);
+	debugfs_create_file("muldf_ip", 0600, dir, p_muldf,
+			    &fops_uldf_ip);
+	debugfs_create_file("muldf_platform", 0600, dir, p_muldf,
+			    &fops_uldf_platform);
+	debugfs_create_file("muldf_driver", 0600, dir, p_muldf,
+			    &fops_uldf_driver);
+	debugfs_create_file("muldf_chip", 0600, dir, p_muldf,
+			    &fops_uldf_chip);
+	debugfs_create_file("muldf_mode", 0600, dir, p_muldf,
+			    &fops_uldf_mode);
+	debugfs_create_file("muldf_module", 0600, dir, p_muldf,
+			    &fops_muldf_module);
+	debugfs_create_file("muldf_dump", 0600, dir, p_muldf,
+			    &fops_muldf_dump);
+	debugfs_create_file("muldf_level", 0600, dir, p_muldf,
+			    &fops_muldf_level);
+	return 0;
+}
diff --git a/net/muldf_helper/muldf_helper.c b/net/muldf_helper/muldf_helper.c
new file mode 100644
index 0000000..3f06de4
--- /dev/null
+++ b/net/muldf_helper/muldf_helper.c
@@ -0,0 +1,977 @@
+// SPDX-License-Identifier: ISC
+/* Copyright (C) 2024 MediaTek Inc. */
+
+#include "muldf_helper.h"
+#include "asm-generic/int-ll64.h"
+#include "asm/string.h"
+#include "linux/fscrypt.h"
+#include "linux/kernel.h"
+#include "linux/list.h"
+#include "linux/printk.h"
+#include "linux/rcupdate.h"
+#include "linux/skbuff.h"
+#include "linux/stddef.h"
+#include "trace.h"
+
+struct muldf_helper *p_muldf;
+
+static u16 checksum16(u8 *buf, int len)
+{
+	int sum = 0;
+
+	while (len > 1) {
+		sum += *((u16 *)buf);
+
+		buf = buf + 2;
+
+		if (sum & 0x80000000)
+			sum = (sum & 0xFFFF) + (sum >> 16);
+		len -= 2;
+	}
+
+	if (len)
+		sum += *((u8 *)buf);
+
+	while (sum >> 16)
+		sum = (sum & 0xFFFF) + (sum >> 16);
+
+	return ~sum;
+}
+
+static inline bool is_str_empty(const char *str)
+{
+	return (!str || (str[0] == '\0'));
+}
+
+static inline bool is_str_valid(const char *str)
+{
+	return !(is_str_empty(str) || strlen(str) >= MAX_NAME_LEN);
+}
+
+static inline void str_empty_handle(char **p_str, char *str)
+{
+	*p_str = is_str_empty(str) ? MULDF_INVALID_STR : str;
+}
+
+int muldf_helper_get_local_mac(void)
+{
+	struct net_device *netdev = NULL;
+	struct net *net = &init_net;
+
+	for_each_netdev(net, netdev) {
+		if (!strcmp(netdev->name, "br-lan")) {
+			memcpy(p_muldf->local_mac, netdev->dev_addr, ETH_ALEN);
+
+			/* fake a source address for transmission */
+			if (p_muldf->local_mac[1] == 0xff)
+				p_muldf->local_mac[1] = 0;
+			else
+				p_muldf->local_mac[1] += 64;
+
+			return 0;
+		}
+	}
+
+	return -1;
+}
+
+static bool muldf_helper_event_status(u16 event_id)
+{
+	struct muldf_event_entry *item;
+	unsigned long flags;
+
+	if (p_muldf->enable == MTK_EVENT_DISABLED)
+		return false;
+
+	if (p_muldf->enable == MTK_EVENT_ENABLED_ALL)
+		return true;
+
+	spin_lock_irqsave(&p_muldf->event_entry_lock, flags);
+	list_for_each_entry(item, &p_muldf->event_entry_head, list) {
+		if (item->id == event_id) {
+			spin_unlock_irqrestore(&p_muldf->event_entry_lock, flags);
+			return true;
+		}
+	}
+	spin_unlock_irqrestore(&p_muldf->event_entry_lock, flags);
+
+	return false;
+}
+
+int muldf_helper_add_event(u16 id)
+{
+	struct muldf_event_entry *entry;
+	unsigned long flags;
+
+	entry = kmalloc(sizeof(*entry), GFP_KERNEL);
+	if (!entry)
+		return -1;
+
+	entry->id = id;
+	INIT_LIST_HEAD(&entry->list);
+
+	spin_lock_irqsave(&p_muldf->event_entry_lock, flags);
+
+	if (p_muldf->event_entry_num > EVENT_ID_MAX_NUM) {
+		spin_unlock_irqrestore(&p_muldf->event_entry_lock, flags);
+		list_del(&entry->list);
+		kfree(entry);
+		return -1;
+	}
+	list_add(&entry->list, &p_muldf->event_entry_head);
+	p_muldf->event_entry_num++;
+
+	spin_unlock_irqrestore(&p_muldf->event_entry_lock, flags);
+
+	return 0;
+}
+
+int muldf_helper_del_event(void)
+{
+	struct muldf_event_entry *item, *tmp;
+	unsigned long flags;
+
+	spin_lock_irqsave(&p_muldf->event_entry_lock, flags);
+	list_for_each_entry_safe(item, tmp, &p_muldf->event_entry_head, list) {
+		list_del(&item->list);
+		kfree(item);
+	}
+	spin_unlock_irqrestore(&p_muldf->event_entry_lock, flags);
+
+	return 0;
+}
+
+static int muldf_helper_enq_event(struct muldf_event_hdr *ev)
+{
+	struct muldf_event_queue *q = &p_muldf->event_queue;
+	unsigned long flags;
+
+	spin_lock_irqsave(&q->lock, flags);
+
+	if ((q->tail + 1) % MAX_EVENT_CNT == q->head) {
+		spin_unlock_irqrestore(&q->lock, flags);
+		return -1;
+	}
+
+	q->events[q->tail] = *ev;
+	q->tail = (q->tail + 1) % MAX_EVENT_CNT;
+	q->cnt++;
+	q->total_cnt++;
+
+	atomic_set(&p_muldf->event_queue_status, 1);
+	wake_up_interruptible(&p_muldf->event_queue_wait);
+
+	spin_unlock_irqrestore(&q->lock, flags);
+
+	return 0;
+}
+
+int muldf_helper_deq_event(struct muldf_event_hdr *ev)
+{
+	struct muldf_event_queue *q = &p_muldf->event_queue;
+	unsigned long flags;
+
+	spin_lock_irqsave(&q->lock, flags);
+
+	if (q->head == q->tail) {
+		spin_unlock_irqrestore(&q->lock, flags);
+		return -1;
+	}
+
+	*ev = q->events[q->head];
+	q->head = (q->head + 1) % MAX_EVENT_CNT;
+	q->cnt--;
+
+	if (q->head == q->tail)
+		atomic_set(&p_muldf->event_queue_status, 0);
+
+	spin_unlock_irqrestore(&q->lock, flags);
+
+	return 0;
+}
+
+struct sk_buff *muldf_helper_alloc(int len)
+{
+	struct sk_buff *skb;
+
+	skb = alloc_skb(len, GFP_KERNEL);
+	if (!skb)
+		return NULL;
+
+	return skb;
+}
+
+static void __muldf_helper_xmit(u16 type, struct sk_buff *data, int data_len)
+{
+	u32 src_ip = 0, dest_ip = ~0;
+	struct pseudo_header *pseudo_h;
+	struct muldf_hdr *dbg_h;
+	struct udphdr *udp_h;
+	struct iphdr *iph;
+	struct ethhdr *eth_h;
+	u8 *hdr;
+
+	// local net_device & remoute server valid check
+	if (p_muldf->server_ip == 0 ||
+	    !p_muldf->vnet)
+		return;
+
+	if (is_zero_ether_addr(p_muldf->local_mac) &&
+	    muldf_helper_get_local_mac())
+		return;
+
+	data->dev = p_muldf->vnet;
+
+	/* unicast fwlog pkt - src ip would be x.x.x.254 */
+	if (p_muldf->server_ip != 0xFFFFFFFF) {
+		dest_ip = p_muldf->server_ip;
+		src_ip = (dest_ip & 0x00FFFFFF) | 0xFE000000;
+	}
+
+	/* prepare mtk_dbg header */
+	dbg_h = (struct muldf_hdr *)skb_push(data, sizeof(struct muldf_hdr));
+	dbg_h->ver = 0x0;
+	dbg_h->type = type;
+
+	/* prepare UDP header */
+	hdr = skb_push(data, sizeof(struct udphdr));
+	udp_h = (struct udphdr *)hdr;
+	udp_h->source = htons(PC_SERVER_MULDFH_LOCAL_PORT);
+	udp_h->dest = htons(p_muldf->server_port);
+	udp_h->len = htons(data_len + sizeof(struct udphdr));
+	udp_h->check = 0;
+
+	hdr = hdr - sizeof(struct pseudo_header);
+	pseudo_h = (struct pseudo_header *)hdr;
+	pseudo_h->saddr = src_ip;
+	pseudo_h->daddr = dest_ip;
+	pseudo_h->len = htons(data_len + sizeof(struct udphdr));
+	pseudo_h->protocol = htons(17);
+	udp_h->check = checksum16(hdr, data_len +
+				  sizeof(struct udphdr) +
+				  sizeof(struct pseudo_header));
+
+	/* prepare IP header */
+	iph = (struct iphdr *)skb_push(data, sizeof(struct iphdr));
+	iph->version = 4;
+	iph->ihl = 5;
+	iph->tos = 0;
+	iph->tot_len = htons(data_len + sizeof(struct iphdr) +
+			     sizeof(struct udphdr));
+	iph->id = 0;
+	iph->frag_off = htons(0x4000);		/* fragmentation flags and offset */
+	iph->ttl = 7;				/* time to live */
+	iph->protocol = 17;			/* protocol UDP */
+	iph->saddr = src_ip;			/* source IP */
+	iph->daddr = dest_ip;			/* destination IP */
+	iph->check = 0;
+	iph->check = checksum16((u8 *)iph, sizeof(struct iphdr)); /* IP checksum */
+
+	/* prepare 802.3 header */
+	eth_h = (struct ethhdr *)skb_push(data, LEN_802_3_H);
+	memcpy(eth_h, p_muldf->server_mac, ETH_ALEN);
+	memcpy(eth_h->h_source, p_muldf->local_mac, ETH_ALEN);
+	eth_h->h_proto = htons(0x0800);
+
+	/* report to upper layer */
+	data->protocol = eth_type_trans(data, data->dev);
+	netif_rx_ni(data);
+}
+
+static void __muldf_helper_xmit_with_hdr(u16 type,
+					 u8 *hdr, u16 hdr_len,
+					 const u8 *data, u16 data_len)
+{
+	struct sk_buff *skb = NULL;
+	/* 802.3 + IP + UDP + dbg */
+	int header_len = LEN_802_3_H + LEN_IPV4_H + LEN_UDP_H
+			 + sizeof(struct muldf_hdr);
+	int len = hdr_len + data_len;
+	bool padding = false;
+	u8 *temp;
+
+	if ((len % 2) != 0) {
+		data_len += 1;
+		len += 1;
+		padding = true;
+	}
+
+	skb = muldf_helper_alloc(header_len + len);
+	if (!skb)
+		return;
+
+	memset(skb->head, 0, header_len + len);
+	skb_reserve(skb, header_len);
+	temp  = skb_put_data(skb, hdr, hdr_len);
+	if (data_len)
+		skb_put_data(skb, data, data_len);
+
+	if (padding)
+		*(temp + len) = 0;
+
+	__muldf_helper_xmit(type, skb, sizeof(struct muldf_hdr) + len);
+}
+
+static struct muldf_module *muldf_helper_get_module(char *module)
+{
+	struct muldf_module *item;
+
+	if (!is_str_valid(module))
+		return NULL;
+
+	rcu_read_lock();
+	list_for_each_entry(item, &p_muldf->module_head, list) {
+		if (!strcmp(item->name, module)) {
+			rcu_read_unlock();
+			return item;
+		}
+	}
+	rcu_read_unlock();
+
+	return NULL;
+}
+
+static struct muldf_cat *muldf_helper_get_cat(struct muldf_module *mod, char *cat)
+{
+	struct muldf_cat *item;
+
+	if (!is_str_valid(cat))
+		return NULL;
+
+	rcu_read_lock();
+	list_for_each_entry(item, &mod->cat_head, list) {
+		if (!strcmp(item->name, cat)) {
+			rcu_read_unlock();
+			return item;
+		}
+	}
+	rcu_read_unlock();
+
+	return NULL;
+}
+
+static struct muldf_sub_cat *muldf_helper_get_sub_cat(struct muldf_cat *cat, char *sub_cat)
+{
+	struct muldf_sub_cat *item;
+
+	if (!is_str_valid(sub_cat))
+		return NULL;
+
+	rcu_read_lock();
+	list_for_each_entry(item, &cat->sub_cat_head, list) {
+		if (!strcmp(item->name, sub_cat)) {
+			rcu_read_unlock();
+			return item;
+		}
+	}
+	rcu_read_unlock();
+
+	return NULL;
+}
+
+static struct muldf_module *muldf_helper_check(char *module, char *cat, char *sub_cat, u8 *dump)
+{
+	struct muldf_module *p_mod;
+	struct muldf_cat *p_cat;
+	struct muldf_sub_cat *p_sub_cat;
+
+	// check module
+	p_mod = muldf_helper_get_module(module);
+	if (!p_mod)
+		return NULL;
+
+	/* we allow module print without category,
+	 * but, we not allow module print without wrong category
+	 */
+	if (is_str_empty(cat))
+		return p_mod;
+
+	p_cat = muldf_helper_get_cat(p_mod, cat);
+	if (!p_cat)
+		return NULL;
+
+	/* we allow module print without sub category,
+	 * but, we not allow module print without wrong sub category
+	 */
+	if (is_str_empty(sub_cat))
+		return p_mod;
+
+	p_sub_cat = muldf_helper_get_sub_cat(p_cat, sub_cat);
+	if (!p_sub_cat)
+		return NULL;
+
+	if (dump)
+		*dump = p_sub_cat->pkt_dump && p_sub_cat->pkt_dump_en;
+
+	return p_mod;
+}
+
+int muldf_helper_set_module_level(char *module, u8 level)
+{
+	struct muldf_module *p_mod = NULL;
+	unsigned long flags;
+
+	if (level > MTK_MSG_ERROR)
+		return -1;
+
+	if (!is_str_valid(module))
+		return -1;
+
+	spin_lock_irqsave(&p_muldf->module_lock, flags);
+	list_for_each_entry(p_mod, &p_muldf->module_head, list) {
+		if (!strcmp(p_mod->name, module)) {
+			p_mod->debug_level = level;
+			spin_unlock_irqrestore(&p_muldf->module_lock, flags);
+			return 0;
+		}
+	}
+	spin_unlock_irqrestore(&p_muldf->module_lock, flags);
+
+	return -1;
+}
+
+int muldf_helper_set_module_cat_dump(char *module, char *cat, char *sub_cat, u8 dump_en)
+{
+	struct muldf_module *p_mod = NULL;
+	struct muldf_cat *p_cat = NULL;
+	struct muldf_sub_cat *p_sub_cat = NULL;
+	unsigned long flags;
+
+	if (!is_str_valid(sub_cat))
+		return -1;
+
+	p_mod = muldf_helper_get_module(module);
+	if (!p_mod)
+		return -1;
+
+	p_cat = muldf_helper_get_cat(p_mod, cat);
+	if (!p_cat)
+		return -1;
+
+	spin_lock_irqsave(&p_muldf->module_lock, flags);
+	list_for_each_entry(p_sub_cat, &p_cat->sub_cat_head, list) {
+		if (!strcmp(p_sub_cat->name, sub_cat)) {
+			if (p_sub_cat->pkt_dump != MTK_DUMP_NONE)
+				p_sub_cat->pkt_dump_en = dump_en;
+
+			spin_unlock_irqrestore(&p_muldf->module_lock, flags);
+			return 0;
+		}
+	}
+	spin_unlock_irqrestore(&p_muldf->module_lock, flags);
+
+	return -1;
+}
+
+int muldf_helper_get_module_info(void)
+{
+	struct sub_cat_t {
+		char name[MAX_NAME_LEN];
+		u8 dump;
+		u8 dump_en;
+	};
+	struct cat_t {
+		char name[MAX_NAME_LEN];
+		u8 sub_cat_num;
+		u8 sub_cat[];
+	};
+	struct module_t {
+		char name[MAX_NAME_LEN];
+		u8 debug_level;
+		u8 cat_num;
+		u8 cat[];
+	};
+
+	struct muldf_module *p_mod;
+	struct muldf_cat *p_cat;
+	struct muldf_sub_cat *p_sub_cat;
+	int module_idx = 0;
+
+	rcu_read_lock();
+	list_for_each_entry(p_mod, &p_muldf->module_head, list) {
+		struct module_t *module = NULL;
+		u16 len = 0, cat_len = 0;
+		u8 *data;
+		struct {
+			u16 module_idx;
+			u16 len;
+		} __packed hdr;
+
+		list_for_each_entry(p_cat, &p_mod->cat_head, list) {
+			len += (sizeof(struct cat_t) +
+			       p_cat->sub_cat_num * sizeof(struct sub_cat_t));
+		}
+
+		len += sizeof(struct module_t);
+		data = kmalloc(len, GFP_KERNEL);
+		if (!data) {
+			rcu_read_unlock();
+			return -1;
+		}
+
+		memset(data, 0, len);
+		module = (struct module_t *)data;
+		strcpy(module->name, p_mod->name);
+		module->debug_level = p_mod->debug_level;
+		module->cat_num = p_mod->cat_num;
+
+		list_for_each_entry(p_cat, &p_mod->cat_head, list) {
+			struct cat_t *cat = (struct cat_t *)(module->cat + cat_len);
+			int k = 0;
+
+			strcpy(cat->name, p_cat->name);
+			cat->sub_cat_num = p_cat->sub_cat_num;
+			cat_len += (sizeof(struct cat_t) +
+				   (p_cat->sub_cat_num * sizeof(struct sub_cat_t)));
+
+			list_for_each_entry(p_sub_cat, &p_cat->sub_cat_head, list) {
+				struct sub_cat_t *sub_cat;
+				u16 sub_cat_len = k * sizeof(struct sub_cat_t);
+
+				sub_cat = (struct sub_cat_t *)(cat->sub_cat + sub_cat_len);
+				strcpy(sub_cat->name, p_sub_cat->name);
+				sub_cat->dump = p_sub_cat->pkt_dump;
+				sub_cat->dump_en = p_sub_cat->pkt_dump_en;
+				k++;
+			}
+		}
+
+		hdr.len = len;
+		hdr.module_idx = module_idx++;
+
+		__muldf_helper_xmit_with_hdr(MTK_PKT_MODULE_INFO, (u8 *)&hdr,
+					     sizeof(hdr), data, len);
+	}
+	rcu_read_unlock();
+
+	return 0;
+}
+
+// API export
+void muldf_xmit_ics(const u8 *data, u16 len)
+{
+	struct muldf_ics_hdr hdr = {0};
+
+	if (p_muldf->mode == MTK_MODE_INTERNAL)
+		return;
+
+	hdr.magic = ICS_DUMP_MAGIC;
+	hdr.len = len;
+
+	__muldf_helper_xmit_with_hdr(MTK_PKT_WLAN_ICS, (u8 *)&hdr, sizeof(hdr), data, len);
+}
+EXPORT_SYMBOL(muldf_xmit_ics);
+
+void muldf_xmit_fwlog(const u8 *data, u16 len)
+{
+	struct muldf_ics_hdr hdr = {0};
+
+	if (p_muldf->mode == MTK_MODE_INTERNAL)
+		return;
+
+	hdr.magic = FWLOG_DUMP_MAGIC;
+	hdr.len = len;
+
+	__muldf_helper_xmit_with_hdr(MTK_PKT_WLAN_FWLOG, (u8 *)&hdr, sizeof(hdr), data, len);
+}
+EXPORT_SYMBOL(muldf_xmit_fwlog);
+
+u64 muldf_alloc_pid(void)
+{
+	p_muldf->pid++;
+	if (!p_muldf->pid)
+		p_muldf->pid++;
+
+	return p_muldf->pid;
+}
+EXPORT_SYMBOL(muldf_alloc_pid);
+
+void muldf_xmit_pkt(u64 pid, const u8 *data, u16 len)
+{
+	struct muldf_pkt_hdr hdr = {0};
+
+	if (p_muldf->mode == MTK_MODE_INTERNAL)
+		return;
+
+	hdr.pid = pid;
+	hdr.len = len;
+	hdr.magic = PKT_DUMP_MAGIC;
+
+	__muldf_helper_xmit_with_hdr(MTK_PKT_RAW_DATA, (u8 *)&hdr, sizeof(hdr), data, len);
+}
+EXPORT_SYMBOL(muldf_xmit_pkt);
+
+void muldf_xmit_event(u16 id)
+{
+	/* we sent the event to userspace,
+	 * so that user filter the special event
+	 */
+	if (p_muldf->mode == MTK_MODE_INTERNAL) {
+		struct muldf_event_hdr event;
+
+		if (!p_muldf->attach)
+			return;
+
+		event.event_id = id;
+		event.magic = EVENT_MAGIC;
+		muldf_helper_enq_event(&event);
+
+	// pc-tool mode, we sent event to pc-tool.
+	} else if (p_muldf->mode == MTK_MODE_PC) {
+		struct sk_buff *skb = NULL;
+		struct muldf_event_hdr hdr = {0};
+		bool enable  = muldf_helper_event_status(id);
+
+		if (enable) {
+			/* 802.3 + IP + UDP + dbg */
+			int header_len = LEN_802_3_H + LEN_IPV4_H + LEN_UDP_H
+					+ sizeof(struct muldf_hdr);
+			int len = sizeof(hdr);
+			bool is_padding = false;
+			u8 *data;
+
+			if ((len % 2) != 0) {
+				len += 1;
+				is_padding = true;
+			}
+
+			skb = muldf_helper_alloc(header_len + len);
+			if (!skb)
+				return;
+
+			memset(skb->head, 0, header_len + len);
+
+			skb_reserve(skb, header_len);
+			/* prepare event header */
+			hdr.event_id = id;
+			hdr.magic = EVENT_MAGIC;
+
+			data  = skb_put_data(skb, &hdr, sizeof(hdr));
+			if (is_padding)
+				*(data + sizeof(hdr)) = 0;
+
+			/* send event */
+			__muldf_helper_xmit(MTK_PKT_EVENT, skb,
+					    sizeof(struct muldf_hdr) + len);
+		}
+	}
+}
+EXPORT_SYMBOL(muldf_xmit_event);
+
+int muldf_add_module(char *module, int def_level)
+{
+	struct muldf_module *p_mod;
+	unsigned long flags;
+
+	if (def_level > MTK_MSG_ERROR)
+		return -1;
+
+	if (!is_str_valid(module))
+		return -1;
+
+	// no resource available, return error
+	if (p_muldf->module_num >= MAX_MODULE)
+		return -1;
+
+	// check whether module already registered
+	p_mod = muldf_helper_get_module(module);
+	if (p_mod)
+		return -1;
+
+	p_mod = kmalloc(sizeof(*p_mod), GFP_KERNEL);
+	if (!p_mod)
+		return -1;
+
+	strncpy(p_mod->name, module, MAX_NAME_LEN);
+	p_mod->debug_level = def_level;
+	p_mod->cat_num = 0;
+	INIT_LIST_HEAD(&p_mod->cat_head);
+	INIT_LIST_HEAD(&p_mod->list);
+
+	spin_lock_irqsave(&p_muldf->module_lock, flags);
+	list_add(&p_mod->list, &p_muldf->module_head);
+	spin_unlock_irqrestore(&p_muldf->module_lock, flags);
+
+	p_muldf->module_num++;
+
+	return 0;
+}
+EXPORT_SYMBOL(muldf_add_module);
+
+int muldf_del_module(char *module)
+{
+	struct muldf_module *p_mod = NULL, *tmp_mod;
+	struct muldf_cat *p_cat, *tmp_cat;
+	struct muldf_sub_cat *p_sub_cat, *tmp_sub_cat;
+
+	if (!is_str_valid(module))
+		return -1;
+
+	rcu_read_lock();
+	list_for_each_entry_safe(p_mod, tmp_mod, &p_muldf->module_head, list) {
+		if (!strcmp(p_mod->name, module)) {
+			list_del(&p_mod->list);
+
+			list_for_each_entry_safe(p_cat, tmp_cat, &p_mod->cat_head, list) {
+				list_del(&p_cat->list);
+
+				list_for_each_entry_safe(p_sub_cat, tmp_sub_cat, &p_cat->sub_cat_head, list) {
+					list_del(&p_sub_cat->list);
+					kfree(p_sub_cat);
+				}
+
+				kfree(p_cat);
+			}
+			kfree(p_mod);
+			break;
+		}
+	}
+	rcu_read_unlock();
+
+	return 0;
+}
+EXPORT_SYMBOL(muldf_del_module);
+
+int muldf_add_cat(char *module, char *cat)
+{
+	struct muldf_module *p_mod;
+	struct muldf_cat *p_cat;
+	unsigned long flags;
+
+	if (!is_str_valid(cat))
+		return -1;
+
+	p_mod = muldf_helper_get_module(module);
+	if (!p_mod)
+		return -1;
+
+	// no resource available, return error
+	if (p_mod->cat_num >= MAX_CATEGORY)
+		return -1;
+
+	// check whether category already added
+	p_cat = muldf_helper_get_cat(p_mod, cat);
+	if (p_cat)
+		return -1;
+
+	p_cat = kmalloc(sizeof(*p_cat), GFP_KERNEL);
+	if (!p_cat)
+		return -1;
+
+	strncpy(p_cat->name, cat, MAX_NAME_LEN);
+	p_cat->sub_cat_num = 0;
+	INIT_LIST_HEAD(&p_cat->sub_cat_head);
+	INIT_LIST_HEAD(&p_cat->list);
+
+	spin_lock_irqsave(&p_muldf->module_lock, flags);
+	list_add(&p_cat->list, &p_mod->cat_head);
+	spin_unlock_irqrestore(&p_muldf->module_lock, flags);
+
+	p_mod->cat_num++;
+
+	return 0;
+}
+EXPORT_SYMBOL(muldf_add_cat);
+
+int muldf_add_sub_cat(char *module, char *cat, char *sub_cat, u8 pkt_dump)
+{
+	struct muldf_module *p_mod;
+	struct muldf_cat *p_cat;
+	struct muldf_sub_cat *p_sub_cat;
+	unsigned long flags;
+
+	if (!is_str_valid(sub_cat))
+		return -1;
+
+	p_mod = muldf_helper_get_module(module);
+	if (!p_mod)
+		return -1;
+
+	p_cat = muldf_helper_get_cat(p_mod, cat);
+	if (!p_cat)
+		return -1;
+
+	// no resource available, return error
+	if (p_cat->sub_cat_num >= MAX_SUB_CATEGORY)
+		return -1;
+
+	// check whether sub category is already added
+	p_sub_cat = muldf_helper_get_sub_cat(p_cat, sub_cat);
+	if (p_sub_cat)
+		return -1;
+
+	p_sub_cat = kmalloc(sizeof(*p_sub_cat), GFP_KERNEL);
+	if (!p_sub_cat)
+		return -1;
+
+	strncpy(p_sub_cat->name, sub_cat, MAX_NAME_LEN);
+	p_sub_cat->pkt_dump = pkt_dump;
+	p_sub_cat->pkt_dump_en = 0;
+	INIT_LIST_HEAD(&p_sub_cat->list);
+
+	spin_lock_irqsave(&p_muldf->module_lock, flags);
+	list_add(&p_sub_cat->list, &p_cat->sub_cat_head);
+	spin_unlock_irqrestore(&p_muldf->module_lock, flags);
+
+	p_cat->sub_cat_num++;
+
+	return 0;
+}
+EXPORT_SYMBOL(muldf_add_sub_cat);
+
+void muldf_print(int level, char *module,
+		 char *cat, char *sub_cat, char *tag,
+		 const char *func, const u32 line,
+		 const char *fmt, ...)
+{
+	va_list args;
+	char *p_cat, *p_sub_cat, *p_tag, cbuf[MULDF_MAX_PRINT_LEN];
+	struct muldf_module *p_mod = muldf_helper_check(module, cat, sub_cat, NULL);
+	u16 cbuf_len;
+	u16 total_len = sizeof(cbuf);
+	u16 prefix_len = 0;
+
+	if (!p_mod)
+		return;
+
+	if (level < p_mod->debug_level)
+		return;
+
+	str_empty_handle(&p_cat, cat);
+	str_empty_handle(&p_sub_cat, sub_cat);
+	str_empty_handle(&p_tag, tag);
+
+	if (func)
+		prefix_len = snprintf(cbuf, total_len, "%s %d:", func, line);
+
+	va_start(args, fmt);
+	vsnprintf(cbuf + prefix_len, total_len - prefix_len, fmt, args);
+	va_end(args);
+
+	cbuf_len = strlen(cbuf) + 1;
+	muldf_helper_print(level, module, 0, p_cat, p_sub_cat, p_tag, cbuf, cbuf_len);
+}
+EXPORT_SYMBOL(muldf_print);
+
+void muldf_print_hex(int level, char *module,
+		     char *cat, char *sub_cat, char *tag,
+		     u8 *buf, u16 len, const char *title)
+{
+	char *p_cat, *p_sub_cat, *p_tag;
+	struct muldf_module *p_mod = muldf_helper_check(module, cat, sub_cat, NULL);
+
+	if (!p_mod)
+		return;
+
+	if (level < p_mod->debug_level)
+		return;
+
+	str_empty_handle(&p_cat, cat);
+	str_empty_handle(&p_sub_cat, sub_cat);
+	str_empty_handle(&p_tag, tag);
+
+	muldf_helper_print_hex(level, module, p_cat, p_sub_cat, p_tag, buf, len, title);
+}
+EXPORT_SYMBOL(muldf_print_hex);
+
+void muldf_print_pkt(int level, char *module,
+		     char *cat, char *sub_cat, char *tag,
+		     u8 *buf, u16 len, const char *fmt, ...)
+{
+	u64 pid;
+	u8 dump = 0;
+	char *p_cat, *p_sub_cat, *p_tag, cbuf[MULDF_MAX_PRINT_LEN];
+	va_list args;
+	struct muldf_module *p_mod = muldf_helper_check(module, cat, sub_cat, &dump);
+	u16 cbuf_len;
+
+	if (!p_mod)
+		return;
+
+	if (level < p_mod->debug_level)
+		return;
+
+	// [key, value] = [skb's address, pid]
+	// TODO: record more info to filter the same packet
+	if (dump) {
+		pid = muldf_alloc_pid();
+
+		str_empty_handle(&p_cat, cat);
+		str_empty_handle(&p_sub_cat, sub_cat);
+		str_empty_handle(&p_tag, tag);
+
+		va_start(args, fmt);
+		vsnprintf(cbuf, sizeof(cbuf), fmt, args);
+		va_end(args);
+
+		cbuf_len = strlen(cbuf) + 1;
+		muldf_helper_print(level, module, pid, p_cat, p_sub_cat, p_tag, cbuf, cbuf_len);
+
+		if (len > 0)
+			muldf_xmit_pkt(pid, buf, len);
+	}
+}
+EXPORT_SYMBOL(muldf_print_pkt);
+
+static int __init muldf_hepler_init(void)
+{
+	int ret = 0;
+	u8 def_mac[ETH_ALEN] = {0x00, 0x01, 0x02, 0x03, 0x04, 0x05};
+
+	p_muldf = kzalloc(sizeof(*p_muldf), GFP_KERNEL);
+	if (!p_muldf)
+		return -ENOMEM;
+
+	memset(p_muldf, 0, sizeof(struct muldf_helper));
+	ret = muldf_helper_vnet_init();
+	if (ret)
+		goto err;
+
+	INIT_LIST_HEAD(&p_muldf->module_head);
+	spin_lock_init(&p_muldf->module_lock);
+
+	INIT_LIST_HEAD(&p_muldf->event_entry_head);
+	spin_lock_init(&p_muldf->event_entry_lock);
+	spin_lock_init(&p_muldf->event_queue.lock);
+	init_waitqueue_head(&p_muldf->event_queue_wait);
+	atomic_set(&p_muldf->event_queue_status, 0);
+
+	// default, internal handle mode & allow event to userspace daemon
+	p_muldf->mode = MTK_MODE_INVALID;
+	scnprintf(p_muldf->platform, sizeof(p_muldf->platform), "mt7987");
+	scnprintf(p_muldf->wlan_driver, sizeof(p_muldf->wlan_driver), "logan");
+	scnprintf(p_muldf->wifi_chip, sizeof(p_muldf->wifi_chip), "mt7993");
+	p_muldf->enable = MTK_EVENT_ENABLED_ALL;
+
+	// default pc-tool server ip & port
+	memcpy(p_muldf->server_mac, def_mac, ETH_ALEN);
+	p_muldf->server_ip = 0x00000000;
+	p_muldf->server_port = PC_SERVER_DEF_PORT;
+	// default local mac, derived form br-lan
+	scnprintf(p_muldf->local_ifname, sizeof(p_muldf->local_ifname), "br-lan");
+
+	ret = muldf_helper_init_debugfs(p_muldf);
+	if (ret)
+		goto err_vnet;
+
+	return 0;
+
+err_vnet:
+	muldf_helper_vnet_exit();
+
+err:
+	kfree(p_muldf);
+	return ret;
+}
+
+static void __exit muldf_hepler_exit(void)
+{
+	muldf_helper_del_event();
+
+	muldf_helper_vnet_exit();
+
+	kfree(p_muldf);
+}
+
+module_init(muldf_hepler_init);
+module_exit(muldf_hepler_exit);
diff --git a/net/muldf_helper/muldf_helper.h b/net/muldf_helper/muldf_helper.h
new file mode 100644
index 0000000..657e154
--- /dev/null
+++ b/net/muldf_helper/muldf_helper.h
@@ -0,0 +1,214 @@
+/* SPDX-License-Identifier: ISC */
+/* Copyright (C) 2024 MediaTek Inc. */
+
+#include <linux/soc/mediatek/muldf.h>
+#include <linux/inet.h>
+#include "linux/printk.h"
+#include "net/tcp.h"
+#include <linux/init.h>
+#include <linux/module.h>
+#include <linux/kernel.h>
+#include <linux/skbuff.h>
+#include <linux/debugfs.h>
+#include <linux/etherdevice.h>
+#include <linux/udp.h>
+#include <linux/ip.h>
+#include <linux/if_ether.h>
+
+#ifndef __MULDF_HELPER_H
+#define __MULDF_HELPER_H
+
+// muldfd ctrl server ip/port
+#define CTRL_SERVER_IP			"127.0.0.1"
+#define CTRL_SERVER_PORT		(PC_SERVER_DEF_PORT + 3)
+
+// muldfd_cli's port for muldfd's ctrl server connect.
+#define CTRL_SERVER_LOCAL_PORT		(PC_SERVER_DEF_PORT + 4)
+
+// muldfd's port for pc-tool server connect.
+#define PC_SERVER_MULDFD_LOCAL_PORT	(PC_SERVER_DEF_PORT + 5)
+
+// mkt_uldf_helper's default ip/port for pc-tool server connect.
+#define PC_SERVER_IP			"192.168.1.2"
+#define PC_SERVER_DEF_PORT		9527
+#define PC_SERVER_DEF_PORT_MULDFD	(PC_SERVER_DEF_PORT + 1)
+#define PC_SERVER_MULDFH_LOCAL_PORT	(PC_SERVER_DEF_PORT + 2)
+
+#define P_802_3_TYPE		2
+#define LEN_802_3_H		14
+#define LEN_UDP_H		8
+#define LEN_IPV4_H		20
+
+#define EVENT_ID_MAX_NUM	512
+#define EVENT_MAGIC		0xabab
+#define PKT_DUMP_MAGIC		0xcdcd
+#define ICS_DUMP_MAGIC		0xefef
+#define FWLOG_DUMP_MAGIC	0xdfdf
+
+#define MAX_NAME_LEN		32
+#define MAX_MODULE		5
+#define MAX_CATEGORY		20
+#define MAX_SUB_CATEGORY	20
+#define MAX_EVENT_CNT		256
+
+// Specially defined event ID for internal muldf-helper use.
+#define EVENT_ID_MODE_INTERNAL	1999
+#define EVENT_ID_MODE_PC	1998
+#define EVENT_ID_DRIVER_INFO	1997
+
+#define MULDF_INVALID_STR	"null"
+
+enum muldf_pkt_type {
+	MTK_PKT_MODULE_INFO = 0xb000,
+	MTK_PKT_EVENT = 0xb001,
+	MTK_PKT_WLAN_ICS,
+	MTK_PKT_RAM_DUMP,
+	MTK_PKT_RAW_DATA,
+	MTK_PKT_KERNEL_LOG,
+	MTK_PKT_SYSTEM_LOG,
+	MTK_PKT_WLAN_FWLOG,
+	MTK_PKT_MCU_COREDUMP
+};
+
+enum muldf_event_en {
+	MTK_EVENT_DISABLED = 0,
+	MTK_EVENT_ENABLED_BY_EID,
+	MTK_EVENT_ENABLED_ALL,
+	_MTK_EVENT_ENABLED_MAX
+};
+
+enum muldf_mode {
+	MTK_MODE_INVALID = 0,
+	MTK_MODE_INTERNAL,
+	MTK_MODE_PC,
+	_MTK_MODE_MAX
+};
+
+struct muldf_hdr {
+	u32 ver; /*0x0*/
+	u16 type;
+	u16 frag_num;
+	u32 rsv[4];
+};
+
+struct muldf_event_hdr {
+	u16 event_id;
+	u16 magic;  /* 0xabab */
+	u32 rsv[4];
+};
+
+struct muldf_ics_hdr {
+	u16 magic;  /* 0xefef */
+	u16 len;
+	u32 rsv[4];
+};
+
+struct muldf_pkt_hdr {
+	u64 pid;
+	u16 magic;  /* 0xcdcd */
+	u16 len;
+	u32 rsv[4];
+};
+
+struct pseudo_header {
+	u32 saddr;
+	u32 daddr;
+	u16 len;
+	u16 protocol;
+};
+
+struct muldf_event_queue {
+	struct muldf_event_hdr events[MAX_EVENT_CNT];
+	int head;
+	int tail;
+	int cnt;
+	int total_cnt;
+	// lock to protect access to the event queue
+	spinlock_t lock;
+};
+
+struct muldf_event_entry {
+	u16 id;
+	u16 enable;
+	struct list_head list;
+};
+
+struct muldf_sub_cat {
+	char name[MAX_NAME_LEN];
+	u8 pkt_dump;
+	u8 pkt_dump_en;
+	struct list_head list;
+};
+
+struct muldf_cat {
+	char name[MAX_NAME_LEN];
+	u8 sub_cat_num;
+	struct list_head sub_cat_head;
+	struct list_head list;
+};
+
+struct muldf_module {
+	char name[MAX_NAME_LEN];
+	u8 debug_level;
+	u8 cat_num;
+	struct list_head cat_head;
+	struct list_head list;
+};
+
+struct muldf_helper {
+	u8 mode;
+	u64 pid;
+	// for common part
+	char platform[16];    // platform/apsoc name: mt7988/mt7986....
+	char wlan_driver[16]; // wifi driver name: logan/mt76/none
+	char wifi_chip[16];   // wifi chip name: mt7986/mt7990/mt7992....
+	char local_ifname[16];
+	u8 local_mac[ETH_ALEN];
+	u8 module_num;
+	// module_lock to protect access to muldf moudle
+	spinlock_t module_lock;
+	struct list_head module_head;
+	struct net_device *vnet;
+	// for pc daemon
+	u32 server_ip;
+	u16 server_port;
+	u8 server_mac[ETH_ALEN];
+	// for local userspace daemon
+	struct muldf_event_queue event_queue;
+	wait_queue_head_t event_queue_wait;
+	atomic_t event_queue_status;
+	char uuid[64];
+	char url[128];
+	u8 mac[ETH_ALEN];
+	u8 attach;
+	// for event info records
+	u8 enable;
+	u16 event_entry_num;
+	struct list_head event_entry_head;
+	// event_entry_lock to protect access to the event entry
+	spinlock_t event_entry_lock;
+};
+
+extern struct muldf_helper *p_muldf;
+
+int muldf_helper_vnet_init(void);
+int muldf_helper_vnet_exit(void);
+
+int muldf_helper_get_local_mac(void);
+int muldf_helper_get_module_info(void);
+int muldf_helper_set_module_level(char *module, u8 level);
+int muldf_helper_set_module_cat_dump(char *module, char *cat, char *sub_cat, u8 dump_en);
+
+int muldf_helper_init_debugfs(struct muldf_helper *p_muldf);
+int muldf_helper_add_event(u16 id);
+int muldf_helper_del_event(void);
+int muldf_helper_deq_event(struct muldf_event_hdr *ev);
+
+void muldf_helper_print(int level, char *module, u64 pid,
+			char *cat, char *sub_cat, char *tag,
+			char *buf, u16 len);
+void muldf_helper_print_hex(int level, char *module,
+			    char *cat, char *sub_cat, char *tag,
+			    u8 *buf, u16 len, const char *title);
+
+#endif /* __MULDF_HELPER_H */
diff --git a/net/muldf_helper/trace.c b/net/muldf_helper/trace.c
new file mode 100644
index 0000000..db2e1d2
--- /dev/null
+++ b/net/muldf_helper/trace.c
@@ -0,0 +1,124 @@
+// SPDX-License-Identifier: ISC
+/* Copyright (C) 2024 MediaTek Inc. */
+
+#include "linux/string.h"
+#include <linux/module.h>
+
+#ifndef __CHECKER__
+#define CREATE_TRACE_POINTS
+#include "trace.h"
+#include "muldf_helper.h"
+
+#define MULDF_DIV_ROUND_UP(n, d) (((n) + (d) - 1) / (d))
+
+#define	MULDF_HEX_FORMAT	" %02x"
+#define	MULDF_HEX_HEAD_FORMAT	"\t\t\t\t\t\t\t\t\t\t\t\t\t\t\t %02x"
+#define	MULDF_HEX_TAIL_FORMAT	" %02x\n"
+#define	MULDF_TIMESTAMP_SIZE	64
+
+static char timestamp[MULDF_TIMESTAMP_SIZE];
+static u64 last_jiffies = 0;
+
+void muldf_helper_get_timestamp(void)
+{
+	struct timespec64 ts;
+	struct tm tm;
+	u64 now = jiffies, interval = HZ;
+	char *weekdays[] = {"Sun", "Mon", "Tue", "Wed", "Thu", "Fri", "Sat"};
+	char *months[] = {"Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"};
+
+	if (time_after(now, last_jiffies + interval)) {
+		ktime_get_real_ts64(&ts);
+		time64_to_tm(ts.tv_sec, 0, &tm);
+
+		snprintf(timestamp, MULDF_TIMESTAMP_SIZE, "%s %s %02d %02d:%02d:%02d %04ld",
+			 weekdays[tm.tm_wday],
+			 months[tm.tm_mon],
+			 tm.tm_mday,
+			 tm.tm_hour, tm.tm_min, tm.tm_sec,
+			 tm.tm_year + 1900);
+
+		last_jiffies = now;
+	}
+}
+
+void muldf_helper_print(int level, char *module, u64 pid,
+			char *cat, char *sub_cat, char *tag,
+			char *buf, u16 len)
+{
+	muldf_helper_get_timestamp();
+
+	if (pid)
+		MULDF_PRINT_PID(timestamp, module, cat, sub_cat, tag, buf, len, pid);
+	else
+		MULDF_PRINT(timestamp, module, cat, sub_cat, tag, buf, len);
+}
+
+void muldf_helper_print_hex(int level, char *module,
+			    char *cat, char *sub_cat, char *tag,
+			    u8 *buf, u16 len, const char *title)
+{
+	u16 i, total_len;
+	const char *display = "";
+	char *strbuf = NULL;
+
+	if (!buf) {
+		display = " [NULL]";
+	} else if (len == 0) {
+		display = "";
+	} else if (len) {
+		u16 head, total, row, row_start, row_end;
+		char strhead[MULDF_MAX_PRINT_LEN];
+
+		snprintf(strhead, sizeof(strhead), "%s - hexdump(len=%d):\n", title, len);
+		head = strlen(strhead);
+		row = MULDF_DIV_ROUND_UP(len, 16);
+
+		/* total lenth calculate:
+		 * head + 1: header string
+		 * len * 3: byte expand to ' %02'
+		 * row: add '\n' on each row
+		 * row * 15: add the '\t' on each row
+		 */
+		total = (head + 1) + (3 * len) + (row) + (row * 15);
+		if (total >= MULDF_MAX_HEX_PRINT_LEN) {
+			display = " [hexdump, len is too large]";
+		} else {
+			strbuf = kmalloc(total, GFP_KERNEL);
+			if (!strbuf)
+				return;
+
+			snprintf(strbuf, (head + 1), "%s", strhead);
+			for (i = 0, row_start = 0, row_end = 0; i < len; i++) {
+				int mod = i % 16, offset;
+
+				offset = head + row_end + i * 3 + (row_start * 15);
+				switch (mod) {
+				case 0:
+					snprintf(&strbuf[offset], 19, MULDF_HEX_HEAD_FORMAT,
+						 buf[i]);
+					row_start++;
+					break;
+				case 15:
+					snprintf(&strbuf[offset], 5, MULDF_HEX_TAIL_FORMAT, buf[i]);
+					row_end++;
+					break;
+				default:
+					snprintf(&strbuf[offset], 4, MULDF_HEX_FORMAT, buf[i]);
+				};
+			}
+
+			display = strbuf;
+		}
+	} else {
+		display = " [REMOVED]";
+	}
+
+	total_len = strlen(display)  + 1;
+	muldf_helper_get_timestamp();
+	MULDF_PRINT_HEX(timestamp, module, cat, sub_cat, tag, display, total_len);
+
+	kfree(strbuf);
+}
+
+#endif
diff --git a/net/muldf_helper/trace.h b/net/muldf_helper/trace.h
new file mode 100644
index 0000000..e0cec07
--- /dev/null
+++ b/net/muldf_helper/trace.h
@@ -0,0 +1,123 @@
+/* SPDX-License-Identifier: ISC */
+/* Copyright (C) 2024 MediaTek Inc. */
+
+#if !defined(__MULDF_TRACE_H__) || defined(TRACE_HEADER_MULTI_READ)
+#define __MULDF_TRACE_H__
+
+#include <linux/tracepoint.h>
+
+#undef TRACE_SYSTEM
+#define TRACE_SYSTEM muldf
+
+#define MULDF_MAX_PRINT_LEN 256
+#define MULDF_MAX_HEX_PRINT_LEN 2048
+
+#define MULDF_PRINT trace_muldfps
+TRACE_EVENT(muldfps,
+	    TP_PROTO(const char *timestamp,
+		     const char *module,
+		     const char *cat,
+		     const char *sub_cat,
+		     const char *tag,
+		     const char *content, u16 len),
+
+	TP_ARGS(timestamp, module, cat, sub_cat, tag, content, len),
+
+	TP_STRUCT__entry(__array(char, timestamp, 64)
+			 __array(char, module, 32)
+			 __array(char, cat, 32)
+			 __array(char, sub_cat, 32)
+			 __array(char, tag, 32)
+			 __dynamic_array(char, content, len)
+	),
+
+	TP_fast_assign(strlcpy(__entry->timestamp, timestamp, 64);
+		       strlcpy(__entry->module, module, 32);
+		       strlcpy(__entry->cat, cat, 32);
+		       strlcpy(__entry->sub_cat, sub_cat, 32);
+		       strlcpy(__entry->tag, tag, 32);
+		       memcpy(__get_dynamic_array(content), content, len);
+	),
+
+	TP_printk("%s: %s: %s: %s: %s: %s", __entry->timestamp,
+		  __entry->module, __entry->cat, __entry->sub_cat,
+		  __entry->tag, __get_dynamic_array(content)
+	)
+);
+
+#define MULDF_PRINT_PID trace_muldfpp
+TRACE_EVENT(muldfpp,
+	    TP_PROTO(const char *timestamp,
+		     const char *module,
+		     const char *cat,
+		     const char *sub_cat,
+		     const char *tag,
+		     const char *content, u16 len, u64 pid),
+
+	TP_ARGS(timestamp, module, cat, sub_cat, tag, content, len, pid),
+
+	TP_STRUCT__entry(__array(char, timestamp, 64)
+			 __array(char, module, 32)
+			 __array(char, cat, 32)
+			 __array(char, sub_cat, 32)
+			 __array(char, tag, 32)
+			 __dynamic_array(char, content, len)
+			 __field(u64, pid)
+	),
+
+	TP_fast_assign(strlcpy(__entry->timestamp, timestamp, 64);
+		       strlcpy(__entry->module, module, 32);
+		       strlcpy(__entry->cat, cat, 32);
+		       strlcpy(__entry->sub_cat, sub_cat, 32);
+		       strlcpy(__entry->tag, tag, 32);
+		       memcpy(__get_dynamic_array(content), content, len);
+		       __entry->pid = pid;
+	),
+
+	TP_printk("%s: %s: %s: %s: %s: (pid=%ld) %s", __entry->timestamp,
+		  __entry->module, __entry->cat, __entry->sub_cat,
+		  __entry->tag, __entry->pid, __get_dynamic_array(content)
+	)
+);
+
+#define MULDF_PRINT_HEX trace_muldfps
+TRACE_EVENT(muldfph,
+	    TP_PROTO(const char *timestamp,
+		     const char *module,
+		     const char *cat,
+		     const char *sub_cat,
+		     const char *tag,
+		     const char *content, u16 len),
+
+	TP_ARGS(timestamp, module, cat, sub_cat, tag, content, len),
+
+	TP_STRUCT__entry(__array(char, timestamp, 64)
+			 __array(char, module, 32)
+			 __array(char, cat, 32)
+			 __array(char, sub_cat, 32)
+			 __array(char, tag, 32)
+			 __dynamic_array(char, content, len)
+	),
+
+	TP_fast_assign(strlcpy(__entry->timestamp, timestamp, 64);
+		       strlcpy(__entry->module, module, 32);
+		       strlcpy(__entry->cat, cat, 32);
+		       strlcpy(__entry->sub_cat, sub_cat, 32);
+		       strlcpy(__entry->tag, tag, 32);
+		       memcpy(__get_dynamic_array(content), content, len);
+	),
+
+	TP_printk("%s: %s: %s: %s: %s: %s", __entry->timestamp,
+		  __entry->module, __entry->cat, __entry->sub_cat,
+		  __entry->tag, __get_dynamic_array(content)
+	)
+);
+
+#endif
+
+#undef TRACE_INCLUDE_PATH
+#define TRACE_INCLUDE_PATH ../../net/muldf_helper
+#undef TRACE_INCLUDE_FILE
+#define TRACE_INCLUDE_FILE trace
+
+#include <trace/define_trace.h>
diff --git a/net/muldf_helper/vnet.c b/net/muldf_helper/vnet.c
new file mode 100644
index 0000000..6b6c374
--- /dev/null
+++ b/net/muldf_helper/vnet.c
@@ -0,0 +1,81 @@
+// SPDX-License-Identifier: ISC
+/* Copyright (C) 2024 MediaTek Inc. */
+
+#include "linux/netdevice.h"
+#include "linux/if_ether.h"
+#include "linux/string.h"
+#include <linux/module.h>
+#include <linux/netdevice.h>
+#include <linux/etherdevice.h>
+#include "linux/printk.h"
+#include "muldf_helper.h"
+
+static int muldf_helper_vnet_open(struct net_device *dev)
+{
+	netif_start_queue(dev);
+
+	return 0;
+}
+
+static int muldf_helper_vnet_stop(struct net_device *dev)
+{
+	netif_stop_queue(dev);
+
+	return 0;
+}
+
+static netdev_tx_t muldf_helper_vnet_start_xmit(struct sk_buff *skb, struct net_device *dev)
+{
+	dev_kfree_skb(skb);
+
+	return NETDEV_TX_OK;
+}
+
+static const struct net_device_ops vnet_netdev_ops = {
+	.ndo_open = muldf_helper_vnet_open,
+	.ndo_stop = muldf_helper_vnet_stop,
+	.ndo_start_xmit = muldf_helper_vnet_start_xmit,
+};
+
+static void muldf_helper_vnet_setup(struct net_device *dev)
+{
+	u8 mac[ETH_ALEN] = {0x00, 0x0a, 0x0b, 0x11, 0x22, 0x33};
+
+	dev->netdev_ops = &vnet_netdev_ops;
+	dev->flags |= IFF_NOARP;
+	dev->features |= NETIF_F_HW_CSUM;
+	ether_setup(dev);
+	memcpy(dev->dev_addr, mac, ETH_ALEN);
+	dev->priv_flags &= ~IFF_TX_SKB_SHARING;
+	dev->features |= NETIF_F_NETNS_LOCAL;
+}
+
+int muldf_helper_vnet_init(void)
+{
+	// create the net_device named vnet
+	p_muldf->vnet = alloc_netdev(0, "vnet0", NET_NAME_USER, muldf_helper_vnet_setup);
+	if (!p_muldf->vnet)
+		return -ENOMEM;
+
+	if (register_netdev(p_muldf->vnet)) {
+		free_netdev(p_muldf->vnet);
+		return -EINVAL;
+	}
+
+	// default active the vnet device
+	netif_carrier_on(p_muldf->vnet);
+	muldf_helper_vnet_open(p_muldf->vnet);
+
+	return 0;
+}
+
+int muldf_helper_vnet_exit(void)
+{
+	muldf_helper_vnet_stop(p_muldf->vnet);
+	netif_carrier_off(p_muldf->vnet);
+
+	unregister_netdev(p_muldf->vnet);
+	free_netdev(p_muldf->vnet);
+
+	return 0;
+}
-- 
2.45.2

